# Exploratory Data Analysis (EDA)  

## Introduction

This notebook presents an Exploratory Data Analysis (EDA) of registered workers from the eShram dataset.

The analysis focuses on understanding the demographic, educational, occupational, and geographic characteristics of the workers.

> **Note:** The current analysis uses a 300-record development sample for exploration and testing. It does not represent the complete eShram dataset.

**Setup**       

In [51]:
import sys
from pathlib import Path

project_root = Path.cwd().parent
sys.path.insert(0, str(project_root))

from ingestion.pipeline import run_pipeline

print("Project root:", project_root)
print("Pipeline imported successfully.")

Project root: c:\Users\ANKAN SEN\OneDrive\Desktop\eshram-labour-intelligence
Pipeline imported successfully.


**Dataset Overview**

In [52]:
df = run_pipeline()


Fetching page 1
Offset: 0
Records received: 100

Fetching page 2
Offset: 100
Records received: 100

Fetching page 3
Offset: 200
Records received: 100

Total records collected: 300

Data Quality Report:
total_rows: 300
total_columns: 12
missing_values: 0
duplicate_rows: 0
invalid_ages: 0
unique_states: 24
unique_districts: 207
unique_occupations: 22
unique_education_levels: 9

Cleaned DataFrame shape: (300, 12)

Data types:
gender                               str
age                                int64
differentlyAbledStatus               str
educationQualification               str
currentStateCode                   int64
currentStateName                     str
currentDistrictCode                int64
currentDistrictName                  str
permanentStateCode                 int64
permanentStateName                   str
primaryOccupation                    str
DataGovUpdateDate         datetime64[us]
dtype: object

Missing values:
gender                    0
age                  

In [53]:
df.head()

,gender,age,differentlyAbledStatus,educationQualification,currentStateCode,currentStateName,currentDistrictCode,currentDistrictName,permanentStateCode,permanentStateName,primaryOccupation,DataGovUpdateDate
0,MALE,34,NOT DISABLED,MIDDLE,19,WEST BENGAL,307,BIRBHUM,19,WEST BENGAL,APPAREL,2024-08-30
1,FEMALE,38,NOT DISABLED,BELOW PRIMARY,9,UTTAR PRADESH,163,HATHRAS,9,UTTAR PRADESH,AGRICULTURE,2024-08-30
2,MALE,56,NOT DISABLED,NOT LITERATE,9,UTTAR PRADESH,127,BALRAMPUR,9,UTTAR PRADESH,ELECTRONICS & HW,2024-08-30
3,MALE,33,NOT DISABLED,SECONDARY,20,JHARKHAND,332,HAZARIBAGH,20,JHARKHAND,AGRICULTURE,2024-08-30
4,FEMALE,30,NOT DISABLED,HIGHER SECONDARY,19,WEST BENGAL,304,24 PARAGANAS SOUTH,19,WEST BENGAL,TOBACCO INDUSTRY,2024-08-30


In [54]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 12 columns):
 #   Column                  Non-Null Count  Dtype         
---  ------                  --------------  -----         
 0   gender                  300 non-null    str           
 1   age                     300 non-null    int64         
 2   differentlyAbledStatus  300 non-null    str           
 3   educationQualification  300 non-null    str           
 4   currentStateCode        300 non-null    int64         
 5   currentStateName        300 non-null    str           
 6   currentDistrictCode     300 non-null    int64         
 7   currentDistrictName     300 non-null    str           
 8   permanentStateCode      300 non-null    int64         
 9   permanentStateName      300 non-null    str           
 10  primaryOccupation       300 non-null    str           
 11  DataGovUpdateDate       300 non-null    datetime64[us]
dtypes: datetime64[us](1), int64(4), str(7)
memory usage: 28.3 KB


**Numerical Distribution**

In [55]:
df['age'].describe()

count    300.000000
mean      35.043333
std       11.495882
min       17.000000
25%       25.000000
50%       33.000000
75%       44.000000
max       59.000000
Name: age, dtype: float64

**Gender Analysis**

In [56]:
gender_counts = df['gender'].value_counts()
gender_counts

gender
MALE      159
FEMALE    141
Name: count, dtype: int64

In [57]:
gender_counts_percentage = (gender_counts*100)/df.shape[0]
gender_counts_percentage

gender
MALE      53.0
FEMALE    47.0
Name: count, dtype: float64

### Gender Distribution

- **Male:** 159 workers (**53%**)
- **Female:** 141 workers (**47%**)
- **Total:** 300 workers

The male representation is **6 percentage points higher** than the female representation.

> **Note:** This analysis is based on the 300-record development sample and should not be generalized to the complete eShram dataset.

**Education Analysis**

In [58]:
Education_counts = df['educationQualification'].value_counts().sort_values(ascending=False)
Education_counts_percentage = (Education_counts*100)/df.shape[0]
print(Education_counts)
print("=============================")
print(Education_counts_percentage)

educationQualification
NOT LITERATE                         71
MIDDLE                               54
HIGHER SECONDARY                     48
SECONDARY                            44
PRIMARY                              39
GRADUATE                             18
BELOW PRIMARY                        16
LITERATE WITHOUT FORMAL SCHOOLING     8
DIPLOMA/CERTIFICATE COURSE            2
Name: count, dtype: int64
educationQualification
NOT LITERATE                         23.666667
MIDDLE                               18.000000
HIGHER SECONDARY                     16.000000
SECONDARY                            14.666667
PRIMARY                              13.000000
GRADUATE                              6.000000
BELOW PRIMARY                         5.333333
LITERATE WITHOUT FORMAL SCHOOLING     2.666667
DIPLOMA/CERTIFICATE COURSE            0.666667
Name: count, dtype: float64


**Education x Gender Analysis**

In [79]:
Education_gender = (
    df.groupby(['educationQualification', 'gender'])
      .size()
      .reset_index(name='count')
)

Education_gender['percentage'] = (
    Education_gender['count']
    / Education_gender.groupby('educationQualification')['count'].transform('sum')
    * 100
).round(2)

Education_gender

,educationQualification,gender,count,percentage
0,BELOW PRIMARY,FEMALE,11,68.75
1,BELOW PRIMARY,MALE,5,31.25
2,DIPLOMA/CERTIFICATE COURSE,FEMALE,1,50.00
3,DIPLOMA/CERTIFICATE COURSE,MALE,1,50.00
4,GRADUATE,FEMALE,10,55.56
5,GRADUATE,MALE,8,44.44
6,HIGHER SECONDARY,FEMALE,21,43.75
7,HIGHER SECONDARY,MALE,27,56.25
8,LITERATE WITHOUT FORMAL SCHOOLING,FEMALE,2,25.00
9,LITERATE WITHOUT FORMAL SCHOOLING,MALE,6,75.00


**Occupation Analysis**

In [105]:
occuption_counts = df['primaryOccupation'].value_counts().head()
occuption_counts

primaryOccupation
AGRICULTURE                       106
CONSTRUCTION                       46
DOMESTIC AND HOUSEHOLD WORKERS     44
APPAREL                            18
AUTOMOBILE & TRANSPORTATION        14
Name: count, dtype: int64

**Occupation concentration**

In [106]:
occupation_counts = (
    df['primaryOccupation'].value_counts() * 100 / df.shape[0]
).round(2)

occupation_counts.head()

primaryOccupation
AGRICULTURE                       35.33
CONSTRUCTION                      15.33
DOMESTIC AND HOUSEHOLD WORKERS    14.67
APPAREL                            6.00
AUTOMOBILE & TRANSPORTATION        4.67
Name: count, dtype: float64

In [110]:
df['primaryOccupation'].value_counts().head().sum()/df.shape[0]*100

# the top 5 occupations reserved 76% of 300-record development sample.

np.float64(76.0)

**Age distribution**

In [116]:
import pandas as pd
age_bins = [16, 24, 34, 44, 54, 59]
age_labels = ['16-24', '25-34', '35-44', '45-54', '55-59']

df['age_group'] = pd.cut(
    df['age'],
    bins=age_bins,
    labels=age_labels,
    include_lowest=True
)

df['age_group'].value_counts().sort_values(ascending=False)

age_group
25-34    91
35-44    70
16-24    66
45-54    58
55-59    15
Name: count, dtype: int64

**Age × Gender Analysis**

In [120]:
Age_gender = (
    df.groupby(['age_group', 'gender'])
      .size()
      .reset_index(name='count')
)

Age_gender

,age_group,gender,count
0,16-24,FEMALE,38
1,16-24,MALE,28
2,25-34,FEMALE,45
3,25-34,MALE,46
4,35-44,FEMALE,26
5,35-44,MALE,44
6,45-54,FEMALE,27
7,45-54,MALE,31
8,55-59,FEMALE,5
9,55-59,MALE,10


**Differently Abled Analysis**


In [123]:
diffAbled_counts = df['differentlyAbledStatus'].value_counts()
diffAbled_percentage = diffAbled_counts/df.shape[0]*100

print(diffAbled_counts)
print(diffAbled_percentage)

differentlyAbledStatus
NOT DISABLED    300
Name: count, dtype: int64
differentlyAbledStatus
NOT DISABLED    100.0
Name: count, dtype: float64


In [124]:
diffAbled_analysis = pd.DataFrame({
    'Status': diffAbled_counts.index,
    'Count': diffAbled_counts.values,
    'Percentage': diffAbled_percentage.values
})

diffAbled_analysis

,Status,Count,Percentage
0,NOT DISABLED,300,100.0


**State-level workforce distribution**

In [140]:
state_workforce_count = df['currentStateName'].value_counts().head()
state_workforce_percentage = (state_workforce_count/df.shape[0]*100).round(2)
state_workforce_percentage

currentStateName
UTTAR PRADESH     30.67
WEST BENGAL       12.00
MADHYA PRADESH     8.00
BIHAR              7.00
RAJASTHAN          6.33
Name: count, dtype: float64

In [141]:
State_Workforce_analysis = pd.DataFrame(
    {
        'currentStateName': state_workforce_count.index,
        'count': state_workforce_count.values,
        'Percentage': state_workforce_percentage.values
    }
)

State_Workforce_analysis 

,currentStateName,count,Percentage
0,UTTAR PRADESH,92,30.67
1,WEST BENGAL,36,12.00
2,MADHYA PRADESH,24,8.00
3,BIHAR,21,7.00
4,RAJASTHAN,19,6.33
